File: minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        if(grid.empty() || grid[0].empty())return 0;
        int R = grid.size()-1,C = grid[0].size()-1;
        for(int i=R-1;i>=0;i--)
            grid[i][C] += grid[i+1][C];
        for(int j=C-1;j>=0;j--)
            grid[R][j] += grid[R][j+1];
        for(int i=R-1;i>=0;i--)
            for(int j=C-1;j>=0;j--)
                grid[i][j] += min(grid[i+1][j],grid[i][j+1]);
        return grid[0][0];
    }
};

File: 30-day-leetcoding-challenge/Minimum Path Sum.java
class Solution {
    private int[][] memo;
    private int[][] grid;
    private int n,m;
    
    private int dp(int i,int j)
    {
        if(i == n-1 && j == m-1)return grid[i][j];
        if(i == n || j == m)return Integer.MAX_VALUE;
        if(memo[i][j] != -1)return memo[i][j];
        return memo[i][j] = grid[i][j] + Math.min(dp(i+1,j),dp(i,j+1));
    }
    
    public int minPathSum(int[][] grid) {
        n = grid.length;
        if(n == 0)return 0;
        m = grid[0].length;
        this.grid = grid;
        memo = new int[n][m];
        for(int i=0;i<n;i++)
            Arrays.fill(memo[i],-1);
        return dp(0,0);
    }
}
